internal/cmd/update.go

```diff
@@ -0,0 +1,309 @@
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"time"
+
+	"cdr.dev/coder-cli/internal/config"
+	"cdr.dev/coder-cli/internal/version"
+	"cdr.dev/coder-cli/pkg/clog"
+	"golang.org/x/xerrors"
+
+	"github.com/blang/semver/v4"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+// updater updates coder-cli.
+type updater struct {
+	confirmF       func(label string) (string, error)
+	executablePath string
+	fs             afero.Fs
+	httpClient     getter
+	versionF       func() string
+}
+
+func updateCmd() *cobra.Command {
+	var (
+		force    bool
+		coderURL string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "Update coder binary",
+		Long:  "Update coder to the version matching a given coder instance.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			httpClient := &http.Client{
+				Timeout: 10 * time.Second,
+			}
+
+			updater := &updater{
+				httpClient:     httpClient,
+				fs:             afero.NewOsFs(),
+				confirmF:       defaultConfirm,
+				versionF:       func() string { return version.Version },
+				executablePath: os.Args[0],
+			}
+			return updater.Run(ctx, force, coderURL)
+		},
+	}
+
+	cmd.Flags().BoolVar(&force, "force", false, "do not prompt for confirmation")
+	cmd.Flags().StringVar(&coderURL, "coder", "", "coder instance against which to match version")
+
+	return cmd
+}
+
+type getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+func (u *updater) Run(ctx context.Context, force bool, coderURLString string) error {
+	// TODO: check under following directories and warn if coder binary is under them:
+	//   * homebrew prefix
+	//   * coder assets root (env CODER_ASSETS_ROOT)
+
+	currentBinaryStat, err := u.fs.Stat(u.executablePath)
+	if err != nil {
+		return clog.Fatal("preflight: cannot stat current binary", clog.Causef("%s", err))
+	}
+
+	if currentBinaryStat.Mode().Perm()&0222 == 0 {
+		return clog.Fatal("preflight: missing write permission on current binary")
+	}
+
+	var coderURL *url.URL
+	if coderURLString == "" {
+		coderURL, err = getCoderConfigURL()
+		if err != nil {
+			return clog.Fatal(
+				"Unable to automatically determine coder URL",
+				clog.Causef(err.Error()),
+				clog.BlankLine,
+				clog.Tipf("use --coder <url> to specify coder URL"),
+			)
+		}
+	} else {
+		coderURL, err = url.Parse(coderURLString)
+		if err != nil {
+			return clog.Fatal("invalid coder URL", err.Error())
+		}
+	}
+
+	desiredVersion, err := getAPIVersionUnauthed(u.httpClient, *coderURL)
+	if err != nil {
+		return clog.Fatal("fetch api version", clog.Causef(err.Error()))
+	}
+
+	clog.LogInfo(fmt.Sprintf("Coder instance at %q reports version %s", coderURL.String(), desiredVersion.String()))
+	clog.LogInfo(fmt.Sprintf("Current version of coder-cli is %s", version.Version))
+
+	if currentVersion, err := semver.Make(u.versionF()); err == nil {
+		if desiredVersion.Compare(currentVersion) == 0 {
```

Comment:
Won't this check also fail if the client is newer than the server? I think that's the correct behavior, but it's not necessarily that the client is newer or up to date with the server, what we are really checking is that the versions match, not necessarily that anything is up-to-date

Reply:
Yes, so it's probably a case of rewording some messages emitted by the tool - "Update to version X-1" sounds very strange.